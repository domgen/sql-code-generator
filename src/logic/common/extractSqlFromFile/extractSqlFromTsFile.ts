import { readFileAsync } from '../readFileAsync';

/*
  note: in order to not block users when they have typescript type errors (by using ts-node),
    we can simply parse the file contents and grab the `export const sql = `__SQL__`;` w/ regex
*/
export const extractSqlFromTsFile = async ({ filePath }: { filePath: string }) => {
  // grab file contents
  const content = await readFileAsync({ filePath });

  // grab the sql definition w/ regex
  const [
    _, // tslint:disable-line no-unused
    sqlMatch,
  ] = new RegExp(/(?:export const sql = `)((.|\n)+)(?:`)/).exec(content) ?? [];

  // throw error if could not extract
  if (!sqlMatch) throw new Error(`could not extract sql from file at path '${filePath}'`);

  // return the sql
  return sqlMatch;
};
