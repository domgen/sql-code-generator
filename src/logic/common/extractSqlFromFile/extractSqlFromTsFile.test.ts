import { extractSqlFromTsFile } from './extractSqlFromTsFile';

const TEST_ASSETS_DIR = `${__dirname}/__test_assets__`;

describe('extractSqlFromTsFile', () => {
  it('should extract sql from a file that only has sql exported', async () => {
    const sql = await extractSqlFromTsFile({
      filePath: `${TEST_ASSETS_DIR}/onlyExportSql.ts`,
    });
    expect(sql).toMatchSnapshot();
  });
  it('should extract sql that exports the sql and imports and exports other things', async () => {
    const sql = await extractSqlFromTsFile({
      filePath: `${TEST_ASSETS_DIR}/importAndExportThingsIncludingSql.ts`,
    });
    expect(sql).toMatchSnapshot();
  });
  it('should throw standard error if sql could not be extracted', async () => {
    try {
      await extractSqlFromTsFile({
        filePath: `${TEST_ASSETS_DIR}/dontExportSql.ts`,
      });
      throw new Error('should not reach here');
    } catch (error) {
      expect(error.message).toContain(
        'could not extract sql from file at path ',
      );
    }
  });
});
