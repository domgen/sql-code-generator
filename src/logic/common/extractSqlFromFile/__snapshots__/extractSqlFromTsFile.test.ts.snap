// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractSqlFromTsFile should extract sql from a file that only has sql exported 1`] = `
"
  select
    u.id,
    concat(first_name, last_name) as full_name
  from user u
  where 1=1
    and name = :name;
"
`;

exports[`extractSqlFromTsFile should extract sql that exports the sql and imports and exports other things 1`] = `
"
  -- query_name = find_suggestion_by_id
  select
    s.id,
    s.uuid,
    s.suggestion_source,
    s.external_id,
    s.suggested_idea_id,
    s.status,
    s.result,
    s.resultant_curated_idea_id
  from view_suggestion_current s
  where s.id = :id
"
`;
